### PR TITLE
Task1

### PR DESCRIPTION
**Pull request by Wonder0Maker**

### CODE REVIEW

File: top_films.py [hunk 1]
```diff
@@ -1,53 +1,154 @@
...
+    return title_rating_info
 
 
-def top_films_of_all_times():
+def top_films_of_all_times(data_frame):
     """
     Find the best films of all times
     """
-    top_films_of_all_times = title_rating_info \
+    top_films_of_all_times = data_frame \
```
Comment (wena77) [inline top_films.py]: don't multiply variables, just overwrite `data_frame`, you can even name it just `df`, that's will be enough
look through your code

File: top_films.py [hunk 2]
```diff
@@ -1,53 +1,154 @@
...
+    return top_actors
+
+
+def title_crew_info(data_frame1, data_frame2):
+    """
+    Create data frame by joining
+    """
+    title_crew_info = data_frame1.join(data_frame2, data_frame1.tconst == data_frame2.tconst, 'inner') \
+        .drop(data_frame2.tconst) \
+        .withColumn('director', f.explode(f.split('directors', ',')))
```
Comment (wena77) [inline top_films.py]: you can overwrite `directors` column

File: top_films.py [hunk 3]
```diff
@@ -1,53 +1,154 @@
...
+
+
+def title_cast_info(data_frame1, data_frame2):
+    """
+    Create data frame by joining
+    """
+    title_cast_info = data_frame1.join(data_frame2, data_frame1.tconst == data_frame2.tconst, 'inner') \
+        .drop(data_frame2.tconst)
+
+    return title_cast_info
```
Comment (wena77) [inline top_films.py]: we don't need a separate function for this, do it in a function with appropriate task

File: top_films.py [hunk 4]
```diff
@@ -1,53 +1,154 @@
...
+
+def title_cast_person(data_frame1, data_frame2):
+    """
+    Create data frame by joining
+    """
+    title_cast_person = data_frame1.join(data_frame2, data_frame1.nconst == data_frame2.nconst, 'inner') \
+        .drop(data_frame2.nconst) \
+        .where(f.col('category').like('act%'))
+
+    return title_cast_person
```
Comment (wena77) [inline top_films.py]: here the same

File: main.py [hunk 5]
```diff
@@ -1,7 +1,24 @@
-from top_films import top_films_of_all_times, top_films_of_last_10years, top_films_of_60s
-from connections import write_csv
+import top_films as tf
```
Comment (wena77) [inline main.py]: better import functions here

File: top_films.py [hunk 6]
```diff
@@ -1,53 +1,133 @@
...
+
 
-rating_info = read_tsv('title.ratings.tsv')
+def prepare_rating_info(left_df, right_df):
+    """
+    Create dataframe about film`s title and rating
+    """
+    df = join_dataframes(left_df, right_df, 'tconst') \
+        .where((f.col('titleType') == 'movie') & (f.col('numVotes') >= 100000)) \
+        .orderBy(f.col('averageRating').desc(), f.col('numVotes').desc())
```
Comment (wena77) [inline top_films.py]: we don't need sorting by `f.col('numVotes').desc()`

File: top_films.py [hunk 7]
```diff
@@ -1,53 +1,133 @@
...
         .where(f.col('startYear').between(1960, 1969))
-    return top_films_of_60s.limit(100)
+    return df.limit(100)
+
+
+def explode_by_genres(df):
+    """
+    Create a ranking dataframe by dividing genres
+    """
+    window_genre = Window.partitionBy('genres').orderBy(f.col('averageRating').desc(), f.col('numVotes').desc())
```
Comment (wena77) [inline top_films.py]: we don't need sorting by `f.col('numVotes').desc()`

File: top_films.py (hunk 6 above)
Comment (Wonder0Maker) [reply top_films.py]: it shows a more accurate rating of films

File: top_films.py [hunk 8]
```diff
@@ -1,53 +1,133 @@
...
+    """
+    Find the best films by genres
+    """
+    df = df.select('tconst', 'primaryTitle', 'startYear', 'genres', 'averageRating', 'numVotes') \
+        .where(f.col('rank_genres') <= 10)
+
+    return df
+
+
+def top_films_by_genres_decades(df):
```
Comment (wena77) [inline top_films.py:73]: this funtion works wrong

File: connections.py [hunk 9]
```diff
@@ -23,4 +23,4 @@ def write_csv(data_frame, file_name):
     """
     data_frame.coalesce(1).write.format('csv') \
         .option('header', True).mode('overwrite') \
-        .save('outputs\outputs{}'.format(file_name))
+        .save('outputs\{}'.format(file_name))
```
Comment (wena77) [inline connections.py]: use `/` not `\`

**Review by wena77 — APPROVED** [review]

**Pull request merged**